module_auto_update/models/module.py

```diff
@@ -1,70 +1,148 @@
+                exclude_patterns,
+                keep_langs,
+            )
+        else:
+            checksum_dir = False
+
+        return checksum_dir
 
     @api.model
-    def create(self, vals):
-        res = super(Module, self).create(vals)
-        res._store_checksum_installed(vals)
-        return res
+    def _get_saved_checksums(self):
+        Icp = self.env['ir.config_parameter']
+        return json.loads(Icp.get_param(PARAM_INSTALLED_CHECKSUMS, '{}'))
 
-    @api.multi
-    def write(self, vals):
-        res = super(Module, self).write(vals)
-        self._store_checksum_installed(vals)
-        return res
+    @api.model
+    def _save_checksums(self, checksums):
+        Icp = self.env['ir.config_parameter']
+        Icp.set_param(PARAM_INSTALLED_CHECKSUMS, json.dumps(checksums))
+
+    @api.model
+    def _save_installed_checksums(self):
+        checksums = {}
+        installed_modules = self.search([('state', '=', 'installed')])
+        for module in installed_modules:
+            checksums[module.name] = module._get_checksum_dir()
+        self._save_checksums(checksums)
+
+    @api.model
+    def _get_modules_partially_installed(self):
+        return self.search([
+            ('state', 'in', ('to install', 'to remove', 'to upgrade')),
+        ])
+
+    @api.model
+    def _get_modules_with_changed_checksum(self):
+        saved_checksums = self._get_saved_checksums()
+        installed_modules = self.search([('state', '=', 'installed')])
+        return installed_modules.filtered(
+            lambda r: r._get_checksum_dir() != saved_checksums.get(r.name),
+        )
+
+    @api.model
+    def upgrade_changed_checksum(self, overwrite_existing_translations=False):
+        """Run an upgrade of the database, upgrading only changed modules.
+
+        Installed modules for which the checksum has changed since the
+        last successful run of this method are marked "to upgrade",
+        then the normal Odoo scheduled upgrade process
+        is launched.
+
+        If there is no module with a changed checksum, and no module in state
+        other than installed, uninstalled, uninstallable, this method does
+        nothing, otherwise the normal Odoo upgrade process is launched.
+
+        After a successful upgrade, the checksums of installed modules are
+        saved.
+
+        In case of error during the upgrade, an exception is raised.
+        If any module remains to upgrade or to uninstall after the upgrade
+        process, an exception is raised as well.
+
+        Note: this method commits the current transaction at each important
+        step, it is therefore not intended to be run as part of a
+        larger transaction.
+        """
+        _logger.info(
+            "Checksum upgrade starting (i18n-overwrite=%s)...",
+            overwrite_existing_translations
+        )
+
+        tools.config['overwrite_existing_translations'] = \
+            overwrite_existing_translations
```

Comment:
This translation option may not work as expected. We are investigating.

Reply:
OK, I'm gonna rebase all my commits in one, with another fix. Do you have push access to this PR? If so, you can push here the fix for that part after that :wink:

Reply:
My last functional tests show that `overwrite_existing_translations` works well. Our doubts came from an external factor. :smiley: